database/buildSchema/43_views_functions_triggers.sql

```diff
@@ -1542,12 +1542,332 @@ $$
+        LEFT JOIN review AS action_review ON action_review.review_assignment_id = reviewer_assignment.id
+        LEFT JOIN review_status_history AS action_review_status_history ON (action_review_status_history.review_id = action_review.id
+                AND action_review_status_history.is_current = TRUE)
+            -- ASSIGNER ACTIONS
+        LEFT JOIN assigned_sections_by_stage_and_level ON (assigned_sections_by_stage_and_level.application_id = app.id
+                AND assigned_sections_by_stage_and_level.stage_id = stage_status.stage_id
+                AND app.outcome = 'PENDING'
+                AND assigned_sections_by_stage_and_level.assigner_id = userId)
+        LEFT JOIN template_section AS assignable_sections ON assignable_sections.template_id = template.id
+        LEFT JOIN allowed_self_assignable_sections(userId) AS allowed_self_assignable_sections ON allowed_self_assignable_sections.review_assignment_id = reviewer_assignment.id
+        LEFT JOIN review_assignment_assigned_sections(userId) as review_assignment_assigned_sections ON review_assignment_assigned_sections.review_assignment_id = reviewer_assignment.id
+    WHERE
+        app.id = applicationId
+    GROUP BY stage_status.stage_id;
+$$
+LANGUAGE sql
+STABLE;
 
--- APPLICATION_LIST_VIEW
--- Aggregated VIEW method of all data required for application list page
--- Requires an empty table as setof return and smart comment to make orderBy work (https://github.com/graphile/graphile-engine/pull/378)
-DROP TABLE IF EXISTS application_list_shape CASCADE;
+-- Function to update assigner/reviewer lists on applications and insert/update
+-- reviewer/assigner actions on application_reviewer_action, by making calls to
+-- the above `single_application_detail` function for the appropriate
+-- user/application combo(s)
+DROP FUNCTION IF EXISTS public.update_application_reviewer_stats CASCADE;
+CREATE OR REPLACE FUNCTION public.update_application_reviewer_stats ()
+    RETURNS TRIGGER
+     -- This allows the function to be run as admin user, otherwise it'll only
+     -- be able to query rows the current user has access to.
+    SECURITY DEFINER
+    AS $$
+
+DECLARE
+    reviewer_ids integer[];
+    rev_id integer;
+    assigner_ids integer[];
+    ass_id integer;
+
+BEGIN
+-- reviewer and assigner lists
+    IF TG_OP != 'INSERT' THEN
+        WITH lists AS (
+                SELECT reviewers, assigners
+                FROM single_application_detail(NEW.application_id)
+            ) 
+        UPDATE public.application
+            SET reviewer_list = (SELECT reviewers FROM lists),
+            assigner_list = (SELECT assigners FROM lists)
+            WHERE id = NEW.application_id;
+    END IF;
+
+    -- Get list of reviewer_ids to update for 
+    IF TG_OP = 'UPDATE' THEN
+        reviewer_ids = ARRAY(
+            SELECT reviewer_id FROM review_assignment
+            WHERE application_id = NEW.application_id
+        );
+    ELSE
```

Comment:
When `review_assignments` are created, it'll run once for each assignment as rows are inserted. But when *updating*, we need to run for all potentially affected reviewers, hence we loop over the list of connected users.